### DISCUSSION TITLE
Is the detail data of custom events intercepted by tiddlywiki?

### DISCUSSION THREAD

**Discussion by oeyoews** [TiddlyWiki/TiddlyWiki5#8422] [Q&A]
```js
    const event = new Event('menubarNavChange', { detail: 'test' });
    window.menubarNav = menubarNav;
    $tw.rootWidget.dispatchEvent(event);
    // window.dispatchEvent(event);
    console.log(event.detail, event);
```

I want to trigger a `menubarNavChange` event and pass detail, but it seems that detail is intercepted by tiddlywiki, and I can't get `detail`.

![image](https://github.com/user-attachments/assets/cb222918-5db5-46f9-bcaa-31a04bab3d9e)

Is it something I did wrong? @Jermolene

**Comment by saqimtiaz** [comment]
Use `CustomEvent()`, example:
`startEl.dispatchEvent(new CustomEvent('swiped', { bubbles: true, cancelable: true, detail: eventData }));`

**Reply by oeyoews** [reply]
`CustomEvent` is indeed effective, but I am a little confused. The mdn documentation does not mention the detail parameter of `Event`, but it can indeed be used in normal web pages.

**Comment by saqimtiaz** [comment]
@oeyoews apologies, I suspect now that I missed a crucial detail in your original post. Are you trying to send a custom DOM event, or a custom widget message?

`widget.dispatchEvent` is meant for use with widget messages and not DOM events, therefore the CustomEvent approach is not appropriate either. See the following for an example of how to send a custom widget message: https://github.com/TiddlyWiki/TiddlyWiki5/blob/9069992163f114a4eec2162c059e5848f2a7b6d9/core/modules/widgets/action-sendmessage.js#L85-L93

**Comment by oeyoews** [comment]
I'm trying to trigger some TiddlyWiki events in js, some of which require a navigator widget, and I didn't know how to handle them, so I tried the motion plugin approach. Actually, using the startup module to add a custom event to pass data is a bad solution here, because the corresponding listener is not always added at the beginning. Just encapsulating it into a function and triggering it when the user clicks is enough.

```js
module.exports = () => {
    function getNavigatorWidget(widget) {
      const child = widget.children?.[0];
      if (child?.parseTreeNode.type == 'navigator') {
        return child;
      }
      return getNavigatorWidget(child);
    }
    const menubarNav = getNavigatorWidget($tw.rootWidget);
    return menubarNav;
}
```
